### PR TITLE
Linear regression SPMD

### PR DESCRIPTION
**Pull request by KulikovNikita**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -458,6 +458,7 @@ def run(self):
         'daal4py.sklearn.model_selection',
         'onedal',
         'onedal.svm',
+        'onedal.spmd',
```
Comment (samir-nasibli) [inline setup.py:461]:
```suggestion
        'onedal.spmd',
        'onedal.spmd.linear_model',
```

File: setup_sklearnex.py [hunk 2]
```diff
@@ -96,6 +96,7 @@
           'sklearnex.metrics',
           'sklearnex.model_selection',
           'sklearnex.neighbors',
+          'sklearnex.spmd',
```
Comment (samir-nasibli) [inline setup_sklearnex.py:99]:
```suggestion
          'sklearnex.spmd',
          'sklearnex.spmd.linear_model',
```

File: sklearnex/ensemble/forest.py [hunk 3]
```diff
@@ -15,4 +15,4 @@
 # limitations under the License.
 #===============================================================================
 
-from daal4py.sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
+from daal4py.sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
```
Comment (samir-nasibli) [inline sklearnex/ensemble/forest.py:18]: not related.

File: onedal/spmd/linear_model/tests/test_linear_regression.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                    last = n_samples
+                else:
+                    last = first + block
+
+                shard = arrays[a][:, first:last]
+                results[a].append(shard)
+
+        return results
+
+    def run_spmd_training(comm, queue, params, X, y):
```
Comment (Alexsandruss) [inline onedal/spmd/linear_model/tests/test_linear_regression.py:60]: `onedal/spmd/linear_model/tests/test_linear_regression.py` and `onedal/linear_model/tests/test_linear_regression.py` have same file name which breaks pytest

File: onedal/spmd/linear_model/tests/test_linear_regression.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+    @pytest.mark.parametrize('queue', get_queues())
+    def test_generated(queue):
+        comm = MPI.COMM_WORLD
+        rank, size = comm.Get_rank(), comm.Get_size()
+        n_samples, n_features = size * 10, size * 11
+
+        params = { 'fit_intercept' : True }
+        X, y = make_regression(n_samples = n_samples, 
+                               n_features = n_features)
```
Comment (Alexsandruss) [inline onedal/spmd/linear_model/tests/test_linear_regression.py:122]: Add random_state for reproducibility

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

File: onedal/spmd/linear_model/tests/test_linear_regression.py (hunk 5 above)
Comment (KulikovNikita) [reply onedal/spmd/linear_model/tests/test_linear_regression.py]: Fixed by deleting. Not clear how to test this functionality

File: examples/sklearnex/spmd_linear_regression.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+from mpi4py import MPI
+
+import dpctl
+
+from numpy.testing import assert_allclose
+
+from onedal.spmd.linear_model import LinearRegression as LinRegSpmd
+
+from sklearn.datasets import make_regression
```
Comment (Alexsandruss) [inline examples/sklearnex/spmd_linear_regression.py:27]: Group imports without too many empty lines

Comment (Alexsandruss) [inline examples/sklearnex/spmd_linear_regression.py:27]: make_regression is not used

File: onedal/spmd/linear_model/linear_model.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#===============================================================================
+
+from abc import ABC
+
+from ...common._spmd_policy import _get_spmd_policy
+
+from onedal.linear_model import LinearRegression as LinearRegression_Batch
```
Comment (Alexsandruss) [inline onedal/spmd/linear_model/linear_model.py:21]: Group imports

File: tests/test_examples_sklearnex.py [hunk 8]
```diff
@@ -53,7 +53,15 @@ def testit(self):
...
-files = [f for f in os.listdir(examples_path) if f.endswith(".py")]
+filters = [lambda x: 'spmd' not in str(x), lambda x: str(x).endswith(".py")]
+
+
+def joint_filter(x):
+    return all([f(x) for f in filters])
+
+
+files = list(filter(joint_filter, os.listdir(examples_path)))
+
```
Comment (Alexsandruss) [inline tests/test_examples_sklearnex.py:64]: Can this work better?
```suggestion
files = [f for f in os.listdir(examples_path) if f.endswith(".py") and 'spmd' not in f]
```

File: examples/sklearnex/spmd_linear_regression.py (hunk 6 above)
Comment (KulikovNikita) [reply examples/sklearnex/spmd_linear_regression.py:27]: Flake 8 will not be satisfied

File: tests/test_examples_sklearnex.py (hunk 8 above)
Comment (KulikovNikita) [reply tests/test_examples_sklearnex.py:64]: rejected

File: onedal/spmd/linear_model/linear_model.py (hunk 7 above)
Comment (KulikovNikita) [reply onedal/spmd/linear_model/linear_model.py]: Fixed

File: examples/sklearnex/spmd_linear_regression.py (hunk 6 above)
Comment (KulikovNikita) [reply examples/sklearnex/spmd_linear_regression.py]: Fixed

Comment (KulikovNikita) [reply examples/sklearnex/spmd_linear_regression.py]: Done.

File: tests/test_examples_sklearnex.py (hunk 8 above)
Comment (Alexsandruss) [reply tests/test_examples_sklearnex.py:64]: This construction is much complicated than just one `and`

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/edb46e11-ea8e-f162-8f78-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
With dpctl: http://intel-ci.intel.com/edb48b75-1a87-f1c3-b0c4-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
On master infra: http://intel-ci.intel.com/edb48bb9-cc22-f100-9e14-a4bf010d0e2e

**Review by napetrov — APPROVED** [review]

**Pull request merged**